File: app/src/main/java/by/tuzau/andrey/rusandroidapp/data/service/FileService.java
package by.tuzau.andrey.rusandroidapp.data.service;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import by.tuzau.andrey.rusandroidapp.data.domain.AbstractFileSystemElement;
import by.tuzau.andrey.rusandroidapp.data.domain.FileElement;
import by.tuzau.andrey.rusandroidapp.data.domain.FolderElement;
import by.tuzau.andrey.rusandroidapp.data.domain.TypeElement;
import by.tuzau.andrey.rusandroidapp.data.service.exception.ServiceException;

/**
 * Created by User on 20.11.2017.
 */

public class FileService {

    private Context context;
    private Map<String, Boolean> mapIsAudioEmpty;


    public FileService(Context context) {
        this.context = context;
    }

    public FolderElement createFolder() throws ServiceException {
        Map<String, FolderElement> mapFolderElement = new HashMap<>();
        mapIsAudioEmpty = new HashMap<>();
        List<String> listAudioPath = getListAudioPath();
        FolderElement lastFolderElement = null;
        for (String path : listAudioPath) {
            String[] listName = path.split(File.separator);
            String pathCurrentFolder = "";
            for (int i = 1; i < listName.length - 1; i++) {
                String name = listName[i];
                pathCurrentFolder += File.separator + name;

                if (!mapFolderElement.containsKey(pathCurrentFolder)) {
                    FolderElement folderElement = new FolderElement(name, pathCurrentFolder);
                    folderElement.setCountAudioFiles(1);
                    if (lastFolderElement != null) {
                        lastFolderElement.addAbstractFileSystemElement(folderElement);
                        folderElement.setRootFolderElement(lastFolderElement);
                    }
                    lastFolderElement = folderElement;
                    mapFolderElement.put(pathCurrentFolder, folderElement);
                    mapIsAudioEmpty.put(pathCurrentFolder, true);
                } else {
                    lastFolderElement = mapFolderElement.get(pathCurrentFolder);
                    lastFolderElement.addCountAudioFiles(1);
                }
            }
            // Adding audio file
            String audioName = listName[listName.length - 1];
            lastFolderElement.addAbstractFileSystemElement(new FileElement(audioName,
                    pathCurrentFolder + File.separator + audioName));
            mapIsAudioEmpty.put(pathCurrentFolder, false);
        }

        FolderElement rootFolderElement = getRootFolderElement(lastFolderElement);
        reduceTreeFolderElement(rootFolderElement);

        rootFolderElement = getRootFolderElement(lastFolderElement);


        return rootFolderElement;
    }

    private FolderElement getRootFolderElement(FolderElement anyFolderElement) {
        while (anyFolderElement.getRootFolderElement() != null) {
            anyFolderElement = anyFolderElement.getRootFolderElement();
        }
        return anyFolderElement;
    }

    private void reduceTreeFolderElement(FolderElement rootFolderElement) {
        // If only one folder inside this folder and there is not any mp3 files
        boolean isAudioEmpty = mapIsAudioEmpty.get(rootFolderElement.getFullPath());
        if (isAudioEmpty) {
            int countFolder = rootFolderElement.getListFileSystemElement().size();
            if (countFolder == 1) {
                // Combine two folder in one
                FolderElement folderElement = (FolderElement) rootFolderElement.getListFileSystemElement().get(0);
                folderElement.setName(rootFolderElement.getName() + File.separator + folderElement.getName());
                FolderElement topFolderElement = rootFolderElement.getRootFolderElement();
                folderElement.setRootFolderElement(topFolderElement);
                if (topFolderElement != null) {
                    topFolderElement.addAbstractFileSystemElement(folderElement);
                    topFolderElement.getListFileSystemElement().remove(rootFolderElement);
                }
            }
        }
        int countItems = rootFolderElement.getListFileSystemElement().size();
        // Getting around all elements into folderElement
        for (int i = 0; i < countItems; i++) {
            AbstractFileSystemElement item = rootFolderElement.getListFileSystemElement().get(i);
            if (item.getTypeElement() == TypeElement.FOLDER) {
                reduceTreeFolderElement((FolderElement) item);
            }
        }
    }

    private List<String> getListAudioPath() {
        List<String> listPath = null;
        ContentResolver resolver = context.getContentResolver();
        // Data for the request
        Uri uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
        String[] projection = {MediaStore.Audio.Media.DATA};
        String selection = MediaStore.Audio.Media.DATA + " like '%.mp3'";
        Cursor cursor = resolver.query(uri, null, selection, null, null);

        if (cursor != null && cursor.moveToFirst()) {
            listPath = new ArrayList<>();
            int pathId = cursor.getColumnIndex(MediaStore.Audio.Media.DATA);
            do {
                listPath.add((String) cursor.getString(pathId));
            } while (cursor.moveToNext());
        }
        return listPath;
    }
}

File: app/src/main/java/by/tuzau/andrey/rusandroidapp/data/service/FileService.kt
package by.tuzau.andrey.rusandroidapp.data.service

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.provider.MediaStore
import android.util.Log
import by.tuzau.andrey.rusandroidapp.data.domain.FileElement
import by.tuzau.andrey.rusandroidapp.data.domain.FolderElement
import by.tuzau.andrey.rusandroidapp.data.domain.TypeElement
import java.io.File

class FileService(private val context: Context) {

    private val mapIsAudioEmpty: MutableMap<String, Boolean> = mutableMapOf()

    fun createFolder() : FolderElement {
        val mapFolderElement: MutableMap<String, FolderElement> = mutableMapOf()
        val listAudioPath = getListAudioPath()
        var lastFolderElement: FolderElement? = null
        for (path in listAudioPath) {
            val listName = path.split(File.separator)
            var pathCurrentFolder: String = ""
            for (i in 1 until listName.size - 1) {
                val name = listName[i]
                pathCurrentFolder += File.separator + name

                if (!mapFolderElement.containsKey(pathCurrentFolder)) {
                    val folderElement = FolderElement(name, pathCurrentFolder)
                    folderElement.countAudioFiles = 1
                    if (lastFolderElement != null) {
                        lastFolderElement.addAbstractFileSystemElement(folderElement)
                        folderElement.rootFolderElement = lastFolderElement
                    }
                    lastFolderElement = folderElement
                    mapFolderElement.put(pathCurrentFolder, folderElement)
                    mapIsAudioEmpty.put(pathCurrentFolder, true)
                } else {
                    lastFolderElement = mapFolderElement[pathCurrentFolder]
                    lastFolderElement?.addCountAudioFiles(1)
                }
            }
            // Adding audio file
            val audioName = listName.last()
            lastFolderElement?.addAbstractFileSystemElement(FileElement(audioName,
                    pathCurrentFolder + File.separator + audioName))
            mapIsAudioEmpty.put(pathCurrentFolder, false)
        }

        var rootFolderElement = getRootFolderElement(lastFolderElement!!)
        reduceTreeFolderElement(rootFolderElement)

        rootFolderElement = getRootFolderElement(lastFolderElement!!)

        return rootFolderElement
    }

    private fun getRootFolderElement(anyFolderElement: FolderElement): FolderElement {
        var folderElement = anyFolderElement
        while (true) {
            folderElement = folderElement.rootFolderElement ?: break
        }
        return folderElement
    }

    private fun reduceTreeFolderElement(rootFolderElement: FolderElement) {
        // If only one folder inside this folder and there is not any mp3 files
        val isAudioEmpty = mapIsAudioEmpty[rootFolderElement.fullPath]
        if (isAudioEmpty != null && isAudioEmpty) {
            val countFolder = rootFolderElement.listElement.size
            if (countFolder == 1) {
                // Combine two folder in one
                val folderElement = rootFolderElement.listElement[0] as FolderElement
                folderElement.name = rootFolderElement.name + File.separator + folderElement.name
                val topFolderElement = rootFolderElement.rootFolderElement
                folderElement.rootFolderElement = topFolderElement
                if (topFolderElement != null) {
                    topFolderElement.addAbstractFileSystemElement(folderElement)
                    topFolderElement.listElement.remove(rootFolderElement)
                }
            }
        }
        val countItems: Int = rootFolderElement.listElement.size
        // Getting around all elements into folderElement
        for (i in 0 until countItems) {
            val item = rootFolderElement.listElement.get(i)
            if (item.typeElement == TypeElement.FOLDER) {
                reduceTreeFolderElement(item as FolderElement)
            }
        }
    }

    private fun getListAudioPath(): List<String> {
        val listPath = mutableListOf<String>()
        val resolver: ContentResolver = context.contentResolver
        // Data for the request
        val uri: Uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
        val selection = MediaStore.Audio.Media.DATA + " like '%.mp3'";
        val cursor = resolver.query(uri, null, selection, null, null)

        if (cursor != null && cursor.moveToFirst()) {
            val pathId = cursor.getColumnIndex(MediaStore.Audio.Media.DATA)
            do {
                listPath.add(cursor.getString(pathId))
            } while (cursor.moveToNext())
        }
        return listPath
    }

}